File: dotnet/Sers/Sers.Core/Sers.Core.Station/Module/SsApiDiscovery/SersValid/SersValidMng.cs
using System.Collections.Generic;
using System.Reflection;
using System.Text.RegularExpressions;
using Newtonsoft.Json.Linq;
using Sers.Core.Extensions;
using Sers.Core.Module.SsApiDiscovery.ApiDesc.Attribute.Valid;
using Sers.Core.Util.SsError;

namespace Sers.Core.Module.SsApiDiscovery.SersValid
{
    public class SersValidMng
    {

        public static List<SsValidation> GetRpcValidationsFromMethod(MethodInfo info)
        {
            var attrs = info.GetCustomAttributes<SsValidationBaseAttribute>();

            var validations = new List<SsValidation>();
            foreach (var attr in attrs)
            {
                attr.GetSsValidation(validations);
            }
            if (validations.Count == 0) return null;
            return validations;
        }

        public enum EValidType
        {
            /// <summary>
            /// { "type":"Equal","value":"Logined"}
            /// </summary>
            Equal,
            /// <summary>
            /// { "type":"Regex","value":"^\\d{11}$"}
            /// </summary>
            Regex,
            /// <summary>
            /// { "type":"Required" }
            /// </summary>
            Required,


            /// <summary>
            /// { "type":"Null"}   always true
            /// </summary>
            Null,

            ///// <summary>
            ///// { "type":"MaxValue","value":11}
            ///// </summary>
            //MaxValue,
            ///// <summary>
            ///// { "type":"MinValue","value":11}
            ///// </summary>
            //MinValue,
            ///// <summary>
            ///// { "type":"Size","min":11,"max":11}
            ///// </summary>
            //Size
        }

        /*
       [   
            SsRegex("^\\d{11}$", ErrorMessage = "手机号格式不正确,不是11位"),
            SsEqual("15012345678", ErrorMessage = "手机号必须为15012345678"),
            SsRequired(ErrorMessage = "手机号不能为空"),


            SsMaxValue(15000000000, ErrorMessage = "手机号格式不正确,不是11位"),
            SsMinValue(10000000000, ErrorMessage = "手机号格式不正确,不是11位"),         
            SsSize(min = 11, max = 11, ErrorMessage = "手机号格式不正确,不是11位")]
      */

        /*
         ssValidations:
       [
        {"path":"user.userType","ssError":{}, "ssValid":{"type":"Equal","value":"Logined"} }		 
       ]
         *
         */

        public static bool Valid(JObject obj, List<SsValidation> validations,   out SsError ssError)
        {
            ssError = null;

            if (null == validations) return true;
            foreach (var validation in validations)
            {
                if (!Valid(obj, validation))
                {
                    ssError = validation.ssError;
                    return false;
                }
            }           
            return true;
        }


 
        static bool Valid(JObject obj, SsValidation validation)
        {
           
            var ssValid = validation.ssValid;

            JToken value= obj.SelectToken(validation.path);


            EValidType validType= EValidType.Null;

            #region get  validType
            try
            {
                validType = ssValid["type"].Value<string>().StringToEnum<EValidType>();
            }
            catch
            {
            }
            #endregion

          
            switch (validType)
            {

                #region (x.1)SsEqual
                case EValidType.Equal:
                {
                    //{ "type":"Equal","value":"Logined"}
                    if (value.IsNull() ||
                        ssValid["value"].ConvertToString() == value.ConvertToString()
                    )
                        return true;
                    break;
                }
                #endregion

                #region (x.2)SsRegex
                case EValidType.Regex:
                    {
                        //{ "type":"Regex","value":"^\\d{11}$"}
                        if (value.IsNull()) return true;

                        Regex regex = new Regex(ssValid["value"].ConvertToString());

                        if (regex.IsMatch(value.ConvertToString()))
                            return true;
                        break;
                    }
                #endregion

                #region (x.3)SsRequired
                case EValidType.Required:
                    {
                        //{ "type":"Required" }
                        if (!value.IsNull()) return true;
                        break;
                    }
                #endregion

                case EValidType.Null:
                    return true;
            }           
            return false;
        }
         


    }
}

File: java/Sers.ServiceStation/src/Sers/Core/Module/SsApiDiscovery/SersValid/SersValidMng.java
package Sers.Core.Module.SsApiDiscovery.SersValid;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import com.google.gson.JsonObject;

import Sers.Core.Module.Serialization.Serialization;
import Sers.Core.Util.SsError.SsError;

public class SersValidMng {
	 public static SsValidation[] getRpcValidationsFromMethod(Method  info)
     {     
		 List<SsValidation> validations;
	 
		 
		 //region (x.1) SsValidation
		 Sers.Core.Module.SsApiDiscovery.ApiDesc.Valid.SsValidation[] valids
		 = info.getAnnotationsByType(Sers.Core.Module.SsApiDiscovery.ApiDesc.Valid.SsValidation.class);
		 
		 validations= Arrays.stream(valids).map( anotation->{
			 SsValidation valid=new SsValidation();
			 
			 valid.path=anotation.path();
			 valid.ssError= new SsError(anotation.errorMessage(),anotation.errorCode());
			 valid.ssValid = Serialization.Instance.deserializeFromString(anotation.ssValid(), JsonObject.class);
			 return valid;			 
		 }).collect(Collectors.toList());
		 //endregion
		 
	 
		 //region (x.2) SsCallerSource
		 Sers.Core.Module.SsApiDiscovery.ApiDesc.Valid.SsCallerSource[] callerSources
		 = info.getAnnotationsByType(Sers.Core.Module.SsApiDiscovery.ApiDesc.Valid.SsCallerSource.class);
 
		 for(Sers.Core.Module.SsApiDiscovery.ApiDesc.Valid.SsCallerSource callerSource : callerSources) {
			 
			 SsValidation valid;
			 
			 valid=new SsValidation();			 
			 valid.path="caller.source";
			 valid.ssError= SsError.Err_NotAllowed;
			 valid.ssValid = Serialization.Instance.deserializeFromString("{\"type\":\"Equal\",\"value\":\""+callerSource.value()+"\"}", JsonObject.class);
			 validations.add(valid);
			 
			 
			 valid=new SsValidation();			 
			 valid.path="caller.source";
			 valid.ssError= SsError.Err_NotAllowed;
			 valid.ssValid = Serialization.Instance.deserializeFromString("{\"type\":\"Required\"}", JsonObject.class);			 
			 validations.add(valid);		 
		 }			 
		 //endregion
		 
		return validations.toArray(new SsValidation[validations.size()]);     
     }
	 
	 
	 
	 
	 
	 
	 
	 
	// region ValidType

	/**
	 * { "type":"Equal","value":"Logined"}
	 */
	public static final String ValidType_Equal = "Equal";

	/**
	 * { "type":"Regex","value":"^\\d{11}$"}
	 */
	public static final String ValidType_Regex = "Regex";

	/**
	 * { "type":"Required" }
	 */
	public static final String ValidType_Required = "Required";
	// endregion

}
